arch/x86_64/mcount-dynamic.c

```diff
@@ -549,12 +675,17 @@ static int patch_code(struct mcount_dynamic_info *mdi, struct mcount_disasm_info
 	 *     0xb: <other instructions>
 	 *
 	 * The trap will emulate a call to the trampoline while in place.
+	 *
+	 * We ensure that every core sees the trap before patching the critical
+	 * zone, by synchronizing the them.
 	 */
 
 	if (register_trap(origin_code_addr, (void *)mdi->trampoline) == -1)
 		return INSTRUMENT_FAILED;
 	((uint8_t *)origin_code_addr)[0] = 0xcc;
 
+	synchronize_all_cores();
```

Comment:
Hmm.. are you calling it for each function?  If so, I'm afraid it'd slow down the whole process.